sydent/db/valsession.py

```diff
@@ -121,6 +121,64 @@ def getTokenSessionById(self, sid):
 
         return None
 
+    def next_link_differs(self, sid, token, next_link):
+        """Check whether the specified session has already been validated with a
+        next_link provided, and if  next_link value
+
+        :param sid: The session ID
+        :type sid: str
+
+        :param token: The validation token
+        :type token: str
+
+        :param next_link: The next_link parameter used in submitting this validation
+        :type next_link: str
+
+        :returns: Whether the provided next_link differs from the provided
+            token's associated next_link. Returns False if the stored next_link
+            value is NULL.
+        :rtype: bool
+        """
+        cur = self.sydent.db.cursor()
+
+        # Check if this session has already been validated
+        s = self.getTokenSessionById(sid)
+        if not s.validated:
+            # This session has not been validated before, we allow it to be
+            return False
+
+        cur.execute("select next_link_used from threepid_token_auths "
+                    "where validationSession = ? and token = ?",
+                    (sid, token))
+
+        row = cur.fetchone()
+        if not row:
+            return False
+        token_next_link = row[0]
+
+        return token_next_link != next_link
+
+    def set_next_link_for_token(self, sid, token, next_link):
+        """Set which next_link was used when using a session token
+
+        :param sid: The session ID
+        :type sid: str
+
+        :param token: The validation token
+        :type token: str
+
+        :param next_link: The next_link parameter used in submitting this validation
+        :type next_link: str
+        """
+        cur = self.sydent.db.cursor()
+
+        cur.execute("update threepid_token_auths "
+                    "set next_link_used = ?"
+                    "where validationSession = ? and token = ?",
+                    (next_link, sid, token))
+
+        self.sydent.db.commit()
```

Comment:
Don't we need to handle the case where the thing fails and we need to rollback?

Reply:
Which thing, the db transaction or validation of client-provided parameters? This should only run once the rest of the endpoint has been successful.

Reply:
The db transaction/cursor?

Reply:
Does that mean wrapping `self.sydent.db.commit()` in a `try:` and doing a `rollback()` if an exception is generated?

If that's the case, shouldn't we wrap every call to `commit()`?

Reply:
I would have thought so, though maybe its less of an issue with sqlite? You can use the fact that the [connection is a context manager](https://docs.python.org/2/library/sqlite3.html#using-the-connection-as-a-context-manager):

```python
with self.sydent.db:
    ...
```

Reply:
Now that's handy!